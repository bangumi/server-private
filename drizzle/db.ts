import type { ExtractTablesWithRelations } from 'drizzle-orm';
import type { MySqlTransaction } from 'drizzle-orm/mysql-core';
import {
  drizzle,
  type MySql2PreparedQueryHKT,
  type MySql2QueryResultHKT,
} from 'drizzle-orm/mysql2';
import * as mysql from 'mysql2/promise';

import config, { developing, testing } from '@app/lib/config.ts';
import { logger } from '@app/lib/logger.ts';

import * as schema from './schema.ts';

const poolConnection = mysql.createPool({
  host: config.mysql.host,
  user: config.mysql.user,
  database: config.mysql.db,
  password: config.mysql.password,
  port: config.mysql.port,
});

export const db = drizzle(poolConnection, {
  schema,
  mode: 'default',
  logger:
    testing || developing
      ? {
          logQuery(query: string, params: unknown[]) {
            logger.trace('query', { query, params });
          },
        }
      : undefined,
});

export * as op from 'drizzle-orm';

export type Txn = MySqlTransaction<
  MySql2QueryResultHKT,
  MySql2PreparedQueryHKT,
  typeof schema,
  ExtractTablesWithRelations<Record<string, unknown>>
>;
