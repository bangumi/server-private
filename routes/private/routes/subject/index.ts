import { Type as t } from '@sinclair/typebox';

import { NeedLoginError } from '@app/lib/auth/index.ts';
import { Security } from '@app/lib/openapi/index.ts';
import * as res from '@app/lib/types/res.ts';
import type { App } from '@app/routes/type.ts';

const SubjectRes = t.Object(
  {
    id: t.Integer(),
  },
  { $id: 'Subject' },
);

// eslint-disable-next-line @typescript-eslint/require-await
export async function setup(app: App) {
  app.addSchema(res.Error);
  app.addSchema(SubjectRes);

  app.get(
    '/subjects/:subjectID',
    {
      schema: {
        summary: '获取未读通知',
        operationId: 'listNotice',
        // tags: [Tag.Wiki],
        params: t.Object({
          subjectID: t.Integer(),
        }),
        security: [{ [Security.CookiesSession]: [] }],
        response: {
          200: t.Ref(SubjectRes),
          401: t.Ref(res.Error, {
            description: '未登录',
            'x-examples': {
              NeedLoginError: {
                value: res.formatError(NeedLoginError('getting notifications')),
              },
            },
          }),
        },
      },
    },
    ({ auth: { login }, params: { subjectID } }) => {
      if (!login) {
        return { id: subjectID };
      }
      return { id: subjectID };
    },
  );
}
