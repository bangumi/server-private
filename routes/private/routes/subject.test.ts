import { beforeEach, describe, expect, test } from 'vitest';

import { db, op } from '@app/drizzle/db.ts';
import * as schema from '@app/drizzle/schema';
import { emptyAuth } from '@app/lib/auth/index.ts';
import { createTestServer } from '@app/tests/utils.ts';

import { setup } from './subject.ts';
import { CommentState } from '@app/lib/topic/type.ts';

describe('subject', () => {
  test('should get subject', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/subjects/12',
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subject episodes', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/subjects/12/episodes',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subject relations', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/subjects/12/relations',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subject characters', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/subjects/12/characters',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subject staffs persons', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/subjects/12/staffs/persons',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subject staffs positions', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/subjects/12/staffs/positions',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subject recs', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/subjects/12/recs',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subject comments', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/subjects/12/comments',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subject reviews', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: '/subjects/184017/reviews',
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });
});

describe('subject topics', () => {
  const testSubjectID = 12;
  const testTopicID = 100;
  const testTopicPostID = 100;
  const testUserID = 382951;
  const testPostID = 101;

  beforeEach(async () => {
    await db
      .delete(schema.chiiSubjectTopics)
      .where(op.eq(schema.chiiSubjectTopics.subjectID, testSubjectID));
    await db.delete(schema.chiiSubjectPosts).where(op.eq(schema.chiiSubjectPosts.mid, testTopicID));
    await db.insert(schema.chiiSubjectTopics).values({
      id: testTopicID,
      subjectID: testSubjectID,
      createdAt: 1462335911,
      updatedAt: 1462335911,
      uid: testUserID,
      title: 'Test Topic',
      state: 0,
      replies: 1,
      display: 1,
    });
    await db.insert(schema.chiiSubjectPosts).values({
      id: testTopicPostID,
      mid: testTopicID,
      uid: testUserID,
      content: 'Test Topic Content',
      related: 0,
      state: 0,
      createdAt: 1462335911,
    });
    await db.insert(schema.chiiSubjectPosts).values({
      id: testPostID,
      mid: testTopicID,
      uid: testUserID,
      content: 'Test Reply',
      related: 0,
      state: 0,
      createdAt: 1462335911,
    });
  });

  test('should get subject topics', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject({
      method: 'get',
      url: `/subjects/${testSubjectID}/topics`,
      query: { limit: '2', offset: '0' },
    });
    expect(res.json()).toMatchSnapshot();
  });

  test('should create new topic', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUserID,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: `/subjects/${testSubjectID}/topics`,
      method: 'post',
      payload: {
        title: 'New Topic',
        content: 'New Content',
        turnstileToken: 'fake-response',
      },
    });

    expect(res.statusCode).toBe(200);
    const { id } = res.json() as { id: number };

    const [topic] = await db
      .select()
      .from(schema.chiiSubjectTopics)
      .where(op.eq(schema.chiiSubjectTopics.id, id));
    expect(topic?.title).toBe('New Topic');

    const [post] = await db
      .select()
      .from(schema.chiiSubjectPosts)
      .where(op.eq(schema.chiiSubjectPosts.mid, id))
      .limit(1);
    expect(post?.content).toBe('New Content');
  });

  test('should edit own topic', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUserID,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: `/subjects/-/topics/${testTopicID}`,
      method: 'put',
      payload: {
        title: 'Updated Title',
        content: 'Updated Content',
      },
    });

    expect(res.statusCode).toBe(200);

    const [topic] = await db
      .select()
      .from(schema.chiiSubjectTopics)
      .where(op.eq(schema.chiiSubjectTopics.id, testTopicID));
    expect(topic?.title).toBe('Updated Title');
  });

  test('should not edit topic by non-owner', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUserID + 1,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: `/subjects/-/topics/${testTopicID}`,
      method: 'put',
      payload: {
        title: 'Unauthorized Update',
        content: 'Unauthorized Content',
      },
    });

    expect(res.statusCode).toBe(401);
    expect(res.json()).toMatchSnapshot();
  });

  test('should get subject post', async () => {
    const app = createTestServer();
    await app.register(setup);
    const res = await app.inject(`/subjects/-/posts/${testPostID}`);
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should create/edit/delete new post', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUserID,
      },
    });
    await app.register(setup);

    // create post
    const createRes = await app.inject({
      url: `/subjects/-/topics/${testTopicID}/replies`,
      method: 'post',
      payload: {
        content: 'New Reply',
        turnstileToken: 'fake-response',
      },
    });
    expect(createRes.statusCode).toBe(200);
    const { id } = createRes.json() as { id: number };
    const [createdPost] = await db
      .select()
      .from(schema.chiiSubjectPosts)
      .where(op.eq(schema.chiiSubjectPosts.id, id));
    expect(createdPost?.content).toBe('New Reply');
    expect(createdPost?.related).toBe(0);

    // update post
    const updateRes = await app.inject({
      url: `/subjects/-/posts/${id}`,
      method: 'put',
      payload: {
        content: 'Updated Reply',
      },
    });
    expect(updateRes.statusCode).toBe(200);
    const [updatedPost] = await db
      .select()
      .from(schema.chiiSubjectPosts)
      .where(op.eq(schema.chiiSubjectPosts.id, id));
    expect(updatedPost?.content).toBe('Updated Reply');

    // delete post
    const deleteRes = await app.inject({
      url: `/subjects/-/posts/${id}`,
      method: 'delete',
    });
    expect(deleteRes.statusCode).toBe(200);
    const [deletedPost] = await db
      .select()
      .from(schema.chiiSubjectPosts)
      .where(op.eq(schema.chiiSubjectPosts.id, id));
    expect(deletedPost?.state).toBe(CommentState.UserDelete);
  });
});
