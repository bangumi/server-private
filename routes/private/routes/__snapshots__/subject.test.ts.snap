// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`subject > should get subject 1`] = `
Object {
  "airtime": Object {
    "date": "2002-04-02",
    "month": 4,
    "weekday": 2,
    "year": 2002,
  },
  "collection": Object {
    "1": 1159,
    "2": 4534,
    "3": 215,
    "4": 463,
    "5": 136,
  },
  "eps": 27,
  "id": 12,
  "images": Object {
    "common": "https://lain.bgm.tv/pic/cover/c/c2/0a/12_aDoa8.jpg",
    "grid": "https://lain.bgm.tv/pic/cover/g/c2/0a/12_aDoa8.jpg",
    "large": "https://lain.bgm.tv/pic/cover/l/c2/0a/12_aDoa8.jpg",
    "medium": "https://lain.bgm.tv/pic/cover/m/c2/0a/12_aDoa8.jpg",
    "small": "https://lain.bgm.tv/pic/cover/s/c2/0a/12_aDoa8.jpg",
  },
  "info": "27话 / 2002年4月2日 / 浅香守生 / CLAMP / 阿部恒",
  "infobox": Array [
    Object {
      "key": "中文名",
      "values": Array [
        Object {
          "v": "人形电脑天使心",
        },
      ],
    },
    Object {
      "key": "别名",
      "values": Array [
        Object {
          "k": "en",
          "v": "Chobits",
        },
      ],
    },
    Object {
      "key": "话数",
      "values": Array [
        Object {
          "v": "27",
        },
      ],
    },
    Object {
      "key": "放送开始",
      "values": Array [
        Object {
          "v": "2002年4月2日",
        },
      ],
    },
    Object {
      "key": "放送星期",
      "values": Array [
        Object {
          "v": "星期二",
        },
      ],
    },
    Object {
      "key": "官方网站",
      "values": Array [
        Object {
          "v": "http://www.tbs.co.jp/chobits",
        },
      ],
    },
    Object {
      "key": "播放电视台",
      "values": Array [
        Object {
          "v": "日本TBS",
        },
      ],
    },
    Object {
      "key": "其他电视台",
      "values": Array [
        Object {
          "v": "",
        },
      ],
    },
    Object {
      "key": "播放结束",
      "values": Array [
        Object {
          "v": "2002年9月24日",
        },
      ],
    },
    Object {
      "key": "其他",
      "values": Array [
        Object {
          "v": "",
        },
      ],
    },
    Object {
      "key": "Copyright",
      "values": Array [
        Object {
          "v": "",
        },
      ],
    },
    Object {
      "key": "原作",
      "values": Array [
        Object {
          "v": "CLAMP",
        },
      ],
    },
    Object {
      "key": "导演",
      "values": Array [
        Object {
          "v": "浅香守生",
        },
      ],
    },
    Object {
      "key": "人物设定",
      "values": Array [
        Object {
          "v": "阿部恒",
        },
      ],
    },
    Object {
      "key": "音响监督",
      "values": Array [
        Object {
          "v": "三間雅文",
        },
      ],
    },
    Object {
      "key": "音乐",
      "values": Array [
        Object {
          "v": "高浪敬太郎",
        },
      ],
    },
    Object {
      "key": "动画制作",
      "values": Array [
        Object {
          "v": "MADHOUSE",
        },
      ],
    },
    Object {
      "key": "录音调整",
      "values": Array [
        Object {
          "v": "内田誠",
        },
      ],
    },
    Object {
      "key": "音效",
      "values": Array [
        Object {
          "v": "小山健二",
        },
      ],
    },
    Object {
      "key": "录音助手",
      "values": Array [
        Object {
          "v": "鳥羽瀬縁",
        },
      ],
    },
    Object {
      "key": "音响助手",
      "values": Array [
        Object {
          "v": "三井友和",
        },
      ],
    },
    Object {
      "key": "文艺",
      "values": Array [
        Object {
          "v": "浦畑達彦",
        },
      ],
    },
    Object {
      "key": "设定制作",
      "values": Array [
        Object {
          "v": "加茂靖子",
        },
      ],
    },
    Object {
      "key": "制片人",
      "values": Array [
        Object {
          "v": "源生哲雄、関戸雄一、小野達矢",
        },
      ],
    },
    Object {
      "key": "制片协力",
      "values": Array [
        Object {
          "v": "パイオニアLDC、ムービック",
        },
      ],
    },
    Object {
      "key": "製作",
      "values": Array [
        Object {
          "v": "TBS、ちょびっツ製作委員会",
        },
      ],
    },
    Object {
      "key": "副导演",
      "values": Array [
        Object {
          "v": "田中洋之（除13）、井上英紀（仅13）",
        },
      ],
    },
  ],
  "locked": false,
  "metaTags": Array [],
  "name": "ちょびっツ",
  "nameCN": "人形电脑天使心",
  "nsfw": false,
  "platform": Object {
    "alias": "tv",
    "enableHeader": true,
    "id": 1,
    "order": 0,
    "type": "TV",
    "typeCN": "TV",
    "wikiTpl": "TVAnime",
  },
  "rating": Object {
    "count": Array [
      4,
      4,
      5,
      13,
      60,
      395,
      987,
      1226,
      369,
      168,
    ],
    "rank": 731,
    "score": 7.57,
    "total": 3231,
  },
  "redirect": 0,
  "series": false,
  "seriesEntry": 0,
  "summary": "在不久的将来,电子技术飞速发展,电脑成为人们生活中不可缺少的一部分.主角的名字是本须和秀树,是个19岁的少年,由于考试失败,来到东京上补习班,过着贫穷潦倒的生活……
到达东京的第一天,他很幸运的在垃圾堆捡到一个人型电脑,一直以来秀树都非常渴望拥有个人电脑.当他抱着她带返公寓后,却不知如何开机,在意想不到的地方找到开关并开启后,故事就此展开
本须和秀树捡到了人型计算机〔唧〕。虽然不晓得她到底是不是〔Chobits〕，但她的身上似乎藏有极大的秘密。看到秀树为了钱而烦恼，唧出去找打工，没想到却找到了危险的工作！为了让秀树开心，唧开始到色情小屋打工。但她在遭到过度激烈的强迫要求之后失控。让周遭计算机因此而强制停摆。
另一方面，秀树发现好友新保与补习班的清水老师有着不可告人的关系……",
  "tags": Array [
    Object {
      "count": 932,
      "name": "CLAMP",
    },
    Object {
      "count": 718,
      "name": "人形电脑天使心",
    },
    Object {
      "count": 397,
      "name": "叽",
    },
    Object {
      "count": 373,
      "name": "chobits",
    },
    Object {
      "count": 333,
      "name": "MADHouse",
    },
    Object {
      "count": 287,
      "name": "TV",
    },
    Object {
      "count": 274,
      "name": "田中理惠",
    },
    Object {
      "count": 261,
      "name": "2002",
    },
    Object {
      "count": 227,
      "name": "小叽",
    },
    Object {
      "count": 186,
      "name": "治愈",
    },
    Object {
      "count": 165,
      "name": "Chobits~初次感动~",
    },
    Object {
      "count": 137,
      "name": "启动键",
    },
    Object {
      "count": 121,
      "name": "漫画改",
    },
    Object {
      "count": 107,
      "name": "萝莉",
    },
    Object {
      "count": 57,
      "name": "恋爱",
    },
    Object {
      "count": 55,
      "name": "浅香守生",
    },
    Object {
      "count": 53,
      "name": "2002年4月",
    },
    Object {
      "count": 51,
      "name": "漫改",
    },
    Object {
      "count": 40,
      "name": "ちぃ",
    },
    Object {
      "count": 35,
      "name": "科幻",
    },
    Object {
      "count": 32,
      "name": "杉田智和",
    },
    Object {
      "count": 19,
      "name": "S田",
    },
    Object {
      "count": 18,
      "name": "2002年",
    },
    Object {
      "count": 13,
      "name": "搞笑",
    },
    Object {
      "count": 13,
      "name": "clamp抢钱",
    },
    Object {
      "count": 12,
      "name": "动画",
    },
    Object {
      "count": 12,
      "name": "童年",
    },
    Object {
      "count": 11,
      "name": "半年番",
    },
    Object {
      "count": 10,
      "name": "萌",
    },
    Object {
      "count": 9,
      "name": "哥特",
    },
  ],
  "type": 2,
  "volumes": 0,
}
`;

exports[`subject > should get subject characters 1`] = `
Object {
  "data": Array [
    Object {
      "actors": Array [],
      "character": Object {
        "comment": 56,
        "id": 32,
        "images": Object {
          "grid": "https://lain.bgm.tv/pic/crt/g/ce/65/32_crt_0g9f9.jpg?r=1606033650",
          "large": "https://lain.bgm.tv/pic/crt/l/ce/65/32_crt_0g9f9.jpg?r=1606033650",
          "medium": "https://lain.bgm.tv/pic/crt/m/ce/65/32_crt_0g9f9.jpg?r=1606033650",
          "small": "https://lain.bgm.tv/pic/crt/s/ce/65/32_crt_0g9f9.jpg?r=1606033650",
        },
        "info": "性别 女 / 生日 4月13日 / 身高 152cm",
        "lock": false,
        "name": "ちぃ",
        "nameCN": "小叽",
        "nsfw": false,
        "role": 1,
      },
      "order": 0,
      "type": 1,
    },
    Object {
      "actors": Array [],
      "character": Object {
        "comment": 10,
        "id": 33,
        "images": Object {
          "grid": "https://lain.bgm.tv/pic/crt/g/18/2b/33_0c5cdcd5072bb1864cdee4d3d6e_yyZGp.jpg",
          "large": "https://lain.bgm.tv/pic/crt/l/18/2b/33_0c5cdcd5072bb1864cdee4d3d6e_yyZGp.jpg",
          "medium": "https://lain.bgm.tv/pic/crt/m/18/2b/33_0c5cdcd5072bb1864cdee4d3d6e_yyZGp.jpg",
          "small": "https://lain.bgm.tv/pic/crt/s/18/2b/33_0c5cdcd5072bb1864cdee4d3d6e_yyZGp.jpg",
        },
        "info": "性别 男 / 生日 9月3日 / 身高 180cm",
        "lock": false,
        "name": "本須和秀樹",
        "nameCN": "本须和秀树",
        "nsfw": false,
        "role": 1,
      },
      "order": 0,
      "type": 1,
    },
  ],
  "total": 15,
}
`;

exports[`subject > should get subject comments 1`] = `
Object {
  "data": Array [],
  "total": 0,
}
`;

exports[`subject > should get subject episodes 1`] = `
Object {
  "data": Array [
    Object {
      "airdate": "2002-04-03",
      "comment": 16,
      "desc": "",
      "disc": 0,
      "duration": "24m",
      "id": 1027,
      "name": "ちぃ 目覚める",
      "nameCN": "叽，觉醒了",
      "sort": 1,
      "subjectID": 12,
      "type": 0,
    },
    Object {
      "airdate": "2002-04-10",
      "comment": 8,
      "desc": "",
      "disc": 0,
      "duration": "24m",
      "id": 1028,
      "name": "ちぃ でかける",
      "nameCN": "叽，出门",
      "sort": 2,
      "subjectID": 12,
      "type": 0,
    },
  ],
  "total": 27,
}
`;

exports[`subject > should get subject recs 1`] = `
Object {
  "data": Array [],
  "total": 0,
}
`;

exports[`subject > should get subject relations 1`] = `
Object {
  "data": Array [
    Object {
      "order": 0,
      "relation": Object {
        "cn": "画集",
        "desc": "",
        "en": "Album",
        "id": 1004,
        "jp": "",
      },
      "subject": Object {
        "id": 11,
        "images": Object {
          "common": "https://lain.bgm.tv/pic/cover/c/65/12/11_bsxG3.jpg",
          "grid": "https://lain.bgm.tv/pic/cover/g/65/12/11_bsxG3.jpg",
          "large": "https://lain.bgm.tv/pic/cover/l/65/12/11_bsxG3.jpg",
          "medium": "https://lain.bgm.tv/pic/cover/m/65/12/11_bsxG3.jpg",
          "small": "https://lain.bgm.tv/pic/cover/s/65/12/11_bsxG3.jpg",
        },
        "info": "2003-03 / CLAMP / 講談社 / 138",
        "locked": false,
        "name": "ちょびっツの「ツ」の字 - Chobits Fan Book",
        "nameCN": "",
        "nsfw": false,
        "rating": Object {
          "count": Array [
            0,
            0,
            0,
            0,
            0,
            1,
            3,
            6,
            1,
            2,
          ],
          "rank": 0,
          "score": 8,
          "total": 13,
        },
        "type": 1,
      },
    },
  ],
  "total": 2,
}
`;

exports[`subject > should get subject reviews 1`] = `
Object {
  "data": Array [
    Object {
      "entry": Object {
        "createdAt": 1680065731,
        "icon": "https://lain.bgm.tv/pic/photo/g/no_photo.png",
        "id": 319484,
        "public": true,
        "replies": 1,
        "summary": "试试",
        "title": "试试",
        "type": 1,
        "uid": 287622,
        "updatedAt": 1680072373,
      },
      "id": 168120,
      "user": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "group": 0,
        "id": 287622,
        "joinedAt": 0,
        "nickname": "nickname 287622",
        "sign": "sing 287622",
        "username": "287622",
      },
    },
  ],
  "total": 4,
}
`;

exports[`subject > should get subject staffs persons 1`] = `
Object {
  "data": Array [
    Object {
      "positions": Array [
        Object {
          "appearEps": "",
          "summary": "",
          "type": Object {
            "cn": "原作",
            "en": "Original Creator/Original Work",
            "id": 1,
            "jp": "",
          },
        },
      ],
      "staff": Object {
        "comment": 59,
        "id": 39,
        "images": Object {
          "grid": "https://lain.bgm.tv/pic/crt/g/d6/7d/39_prsn_anidb.jpg",
          "large": "https://lain.bgm.tv/pic/crt/l/d6/7d/39_prsn_anidb.jpg",
          "medium": "https://lain.bgm.tv/pic/crt/m/d6/7d/39_prsn_anidb.jpg",
          "small": "https://lain.bgm.tv/pic/crt/s/d6/7d/39_prsn_anidb.jpg",
        },
        "info": "",
        "lock": false,
        "name": "CLAMP",
        "nameCN": "",
        "nsfw": false,
        "type": 3,
      },
    },
    Object {
      "positions": Array [
        Object {
          "appearEps": "",
          "summary": "",
          "type": Object {
            "cn": "导演",
            "en": "Director/Direction",
            "id": 2,
            "jp": "監督 シリーズ監督",
          },
        },
        Object {
          "appearEps": "",
          "summary": "",
          "type": Object {
            "cn": "分镜",
            "en": "Storyboard",
            "id": 4,
            "jp": "コンテ  ストーリーボード  画コンテ  絵コンテ",
          },
        },
        Object {
          "appearEps": "",
          "summary": "",
          "type": Object {
            "cn": "演出",
            "en": "Episode Direction",
            "id": 5,
            "jp": "",
          },
        },
      ],
      "staff": Object {
        "comment": 100,
        "id": 40,
        "images": Object {
          "grid": "https://lain.bgm.tv/pic/crt/g/1e/ef/40_prsn_JQ7B0.jpg?r=1701577139",
          "large": "https://lain.bgm.tv/pic/crt/l/1e/ef/40_prsn_JQ7B0.jpg?r=1701577139",
          "medium": "https://lain.bgm.tv/pic/crt/m/1e/ef/40_prsn_JQ7B0.jpg?r=1701577139",
          "small": "https://lain.bgm.tv/pic/crt/s/1e/ef/40_prsn_JQ7B0.jpg?r=1701577139",
        },
        "info": "性别 男 / 生日 1967-03-11",
        "lock": false,
        "name": "浅香守生",
        "nameCN": "浅香守生",
        "nsfw": false,
        "type": 1,
      },
    },
  ],
  "total": 91,
}
`;

exports[`subject > should get subject staffs positions 1`] = `
Object {
  "data": Array [
    Object {
      "position": Object {
        "cn": "导演",
        "en": "Director/Direction",
        "id": 2,
        "jp": "監督 シリーズ監督",
      },
      "staffs": Array [
        Object {
          "appearEps": "",
          "person": Object {
            "comment": 59,
            "id": 39,
            "images": Object {
              "grid": "https://lain.bgm.tv/pic/crt/g/d6/7d/39_prsn_anidb.jpg",
              "large": "https://lain.bgm.tv/pic/crt/l/d6/7d/39_prsn_anidb.jpg",
              "medium": "https://lain.bgm.tv/pic/crt/m/d6/7d/39_prsn_anidb.jpg",
              "small": "https://lain.bgm.tv/pic/crt/s/d6/7d/39_prsn_anidb.jpg",
            },
            "info": "",
            "lock": false,
            "name": "CLAMP",
            "nameCN": "",
            "nsfw": false,
            "type": 3,
          },
          "summary": "",
        },
      ],
    },
  ],
  "total": 34,
}
`;

exports[`subject topics > should get recent subject topics 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": 1216020847,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "group": 11,
        "id": 2,
        "joinedAt": 0,
        "nickname": "nickname 2",
        "sign": "sing 2",
        "username": "2",
      },
      "creatorID": 2,
      "display": 1,
      "id": 1,
      "parentID": 1,
      "replies": Array [],
      "replyCount": 76,
      "state": 0,
      "subject": Object {
        "id": 1,
        "images": Object {
          "common": "https://lain.bgm.tv/pic/cover/c/c4/ca/1_EcYzQ.jpg",
          "grid": "https://lain.bgm.tv/pic/cover/g/c4/ca/1_EcYzQ.jpg",
          "large": "https://lain.bgm.tv/pic/cover/l/c4/ca/1_EcYzQ.jpg",
          "medium": "https://lain.bgm.tv/pic/cover/m/c4/ca/1_EcYzQ.jpg",
          "small": "https://lain.bgm.tv/pic/cover/s/c4/ca/1_EcYzQ.jpg",
        },
        "info": "1998-09-25 / 蔡智恒 / 紅色文化、知识出版社 / 188",
        "locked": false,
        "name": "第一次的親密接觸",
        "nameCN": "第一次的亲密接触",
        "nsfw": false,
        "rating": Object {
          "count": Array [
            0,
            0,
            1,
            0,
            0,
            9,
            16,
            31,
            8,
            4,
          ],
          "rank": 1573,
          "score": 7.67,
          "total": 69,
        },
        "type": 1,
      },
      "title": "拿这个来测试",
      "updatedAt": 1639999129,
    },
    Object {
      "createdAt": 1628589969,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "group": 0,
        "id": 617229,
        "joinedAt": 0,
        "nickname": "nickname 617229",
        "sign": "sing 617229",
        "username": "617229",
      },
      "creatorID": 617229,
      "display": 1,
      "id": 19887,
      "parentID": 8,
      "replies": Array [],
      "replyCount": 19,
      "state": 0,
      "subject": Object {
        "id": 8,
        "images": Object {
          "common": "https://lain.bgm.tv/pic/cover/c/c9/f0/8_wK0z3.jpg",
          "grid": "https://lain.bgm.tv/pic/cover/g/c9/f0/8_wK0z3.jpg",
          "large": "https://lain.bgm.tv/pic/cover/l/c9/f0/8_wK0z3.jpg",
          "medium": "https://lain.bgm.tv/pic/cover/m/c9/f0/8_wK0z3.jpg",
          "small": "https://lain.bgm.tv/pic/cover/s/c9/f0/8_wK0z3.jpg",
        },
        "info": "25话 / 2008年4月6日 / 谷口悟朗 / 木村貴宏",
        "locked": false,
        "name": "コードギアス 反逆のルルーシュR2",
        "nameCN": "Code Geass 反叛的鲁路修R2",
        "nsfw": false,
        "rating": Object {
          "count": Array [
            44,
            15,
            32,
            66,
            145,
            457,
            1472,
            3190,
            2640,
            1377,
          ],
          "rank": 121,
          "score": 8.19,
          "total": 9438,
        },
        "type": 2,
      },
      "title": "请教大神R2为什么节奏没有R1好",
      "updatedAt": 1639381921,
    },
  ],
  "total": 34,
}
`;

exports[`subject topics > should get subject post 1`] = `
Object {
  "content": "Test Reply",
  "createdAt": 1462335911,
  "creator": Object {
    "avatar": Object {
      "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
      "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
      "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
    },
    "group": 10,
    "id": 382951,
    "joinedAt": 1512262276,
    "nickname": "树洞酱",
    "sign": "treeholechan@gmail.com 密码:lovemeplease",
    "username": "382951",
  },
  "creatorID": 382951,
  "id": 101,
  "state": 0,
  "topic": Object {
    "createdAt": 1462335911,
    "creator": Object {
      "avatar": Object {
        "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
        "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
        "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
      },
      "group": 10,
      "id": 382951,
      "joinedAt": 1512262276,
      "nickname": "树洞酱",
      "sign": "treeholechan@gmail.com 密码:lovemeplease",
      "username": "382951",
    },
    "creatorID": 382951,
    "display": 1,
    "id": 100,
    "parentID": 12,
    "replyCount": 1,
    "state": 0,
    "title": "Test Topic",
    "updatedAt": 1462335911,
  },
}
`;

exports[`subject topics > should get subject topics 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": 1462335911,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
          "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
          "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
        },
        "group": 10,
        "id": 382951,
        "joinedAt": 1512262276,
        "nickname": "树洞酱",
        "sign": "treeholechan@gmail.com 密码:lovemeplease",
        "username": "382951",
      },
      "creatorID": 382951,
      "display": 1,
      "id": 100,
      "parentID": 12,
      "replyCount": 1,
      "state": 0,
      "title": "Test Topic",
      "updatedAt": 1462335911,
    },
  ],
  "total": 1,
}
`;

exports[`subject topics > should not edit topic by non-owner 1`] = `
Object {
  "code": "NOT_ALLOWED",
  "error": "Forbidden",
  "message": "you don't have permission to update topic",
  "statusCode": 403,
}
`;
